src/libstd/sys/windows/mutex.rs

```diff
@@ -157,34 +157,42 @@ fn kind() -> Kind {
     return ret;
 }
 
-pub struct ReentrantMutex { inner: UnsafeCell<c::CRITICAL_SECTION> }
+pub struct ReentrantMutex { inner: UnsafeCell<MaybeUninit<c::CRITICAL_SECTION>> }
 
 unsafe impl Send for ReentrantMutex {}
 unsafe impl Sync for ReentrantMutex {}
 
 impl ReentrantMutex {
-    pub unsafe fn uninitialized() -> ReentrantMutex {
-        mem::uninitialized()
+    pub fn uninitialized() -> ReentrantMutex {
+        ReentrantMutex { inner: UnsafeCell::new(MaybeUninit::uninitialized()) }
     }
 
     pub unsafe fn init(&mut self) {
-        c::InitializeCriticalSection(self.inner.get());
+        c::InitializeCriticalSection((&mut *self.inner.get()).as_mut_ptr());
     }
 
     pub unsafe fn lock(&self) {
-        c::EnterCriticalSection(self.inner.get());
+        // `init` must have been called, so this is now initialized and
+        // we can call `get_mut`.
+        c::EnterCriticalSection((&mut *self.inner.get()).get_mut());
     }
 
     #[inline]
     pub unsafe fn try_lock(&self) -> bool {
-        c::TryEnterCriticalSection(self.inner.get()) != 0
+        // `init` must have been called, so this is now initialized and
+        // we can call `get_mut`.
+        c::TryEnterCriticalSection((&mut *self.inner.get()).get_mut()) != 0
     }
 
     pub unsafe fn unlock(&self) {
-        c::LeaveCriticalSection(self.inner.get());
+        // `init` must have been called, so this is now initialized and
+        // we can call `get_mut`.
+        c::LeaveCriticalSection((&mut *self.inner.get()).get_mut());
     }
 
     pub unsafe fn destroy(&self) {
-        c::DeleteCriticalSection(self.inner.get());
+        // `init` must have been called, so this is now initialized and
+        // we can call `get_mut`.
+        c::DeleteCriticalSection((&mut *self.inner.get()).get_mut());
```

Comment:
These `get_mut` calls could be `as_mut_ptr`, right? This avoids the unresolved question of uninit `&mut T`.

Reply:
It is incorrect to call any of these methods without previously having initialized the mutex. So I don't think the question even comes up.

But you are also right that there is no reason to use `&mut` at all here, so I will change this to `*mut`.